typing/parmatch.ml

```diff
@@ -1007,18 +1007,18 @@ let build_other ext env =
             0 succ d env
       | Constant Const_int32 _ ->
           build_other_constant
-            (function Constant(Const_int32 i) -> i | _ -> assert false)
-            (function i -> Tpat_constant(Const_int32 i))
+            (function Constant(Const_int32 (i, _)) -> i | _ -> assert false)
+            (function i -> Tpat_constant(Const_int32 (i, Int32.to_string i)))
```

Comment:
I don't know how "hot" this function is. If it is "hot", then it may be worth considering changing `Const_int32` from `Const_int32 of int32 * string` to `Const_int32 of int32 * string option` and using `None` here instead of `Int32.to_string` to avoid the overhead of conversion. This would complicate handling of `Const_int32` in some other places.

(Same for `Const_int64` and `Const_nativeint`.)

TODO: Run a benchmark to see if this is an issue at all.

Reply:
This is not really about performance; this function is only called once an appropriate candidate counter-example is found. (In some case this part of the space search may be reached several times due to GADT troubles occurring earlier, but still I think you would be hard-pressed to find a counter-example where the performance actually matter.) On the other hand, `None` has the benefit of providing an easy way out for constants that are produced by code, instead of parsed from user input, and this may be convenient in other places at well. You would know better, as you have a good grasp of this slice of functionality right now.

Reply:
Thanks @gasche! For future reference: the function

```ocaml
(function i -> Tpat_constant(Const_int32 (i, Int32.to_string i)))
```

is passed as the `make` argument to `build_other_constant`, and `make` is only called once the counter-example has been established. I'm no longer worried about performance in this instance.

> On the other hand, `None` has the benefit of providing an easy way out for constants that are produced by code, instead of parsed from user input, and this may be convenient in other places at well.

This makes me wonder where the string comes from in contexts where there is no source code, e.g. in `dumpobj`. So far in this PR, this place in `parmatch.ml` is the only place where we have to synthesise the string representation from the integer value. I would have expected there to be other places (e.g. in `dumpobj`) where this needs to happen as well, but haven't found any.

If there were more places where we would have to "fake" the original string using something like `Int32.to_string` then that would strengthen the case for `string option`.

Reply:
(For reference: https://github.com/ocaml/ocaml/pull/10767#discussion_r748885720 explains more of what is going on in `build_other` generally and why synthesising a decimal is a sensible choice here.)